src/Microsoft.ML.Tokenizers/Model/SentencePieceBaseModel.cs

```diff
@@ -0,0 +1,753 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Sentencepiece;
+using System;
+using System.Buffers;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Text;
+using System.Text.RegularExpressions;
+
+namespace Microsoft.ML.Tokenizers
+{
+    internal abstract class SentencePieceBaseModel
+    {
+        internal SentencePieceBaseModel(ModelProto modelProto, bool addBos = false, bool addEos = false, IReadOnlyDictionary<string, int>? specialTokens = null)
+        {
+            if (modelProto is null)
+            {
+                throw new ArgumentNullException(nameof(modelProto));
+            }
```

Comment:
Maybe we could do more validation of modelProto here.

Reply:
This is answered in the previous comment.